vscode/src/azure/treeRefresher.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { log } from "qsharp-lang";
+import { WorkspaceConnection, WorkspaceTreeProvider } from "./treeView";
+
+// To ensure that for a given workspace, we don't have multiple refreshes running
+const runningTimeouts = new Map<string, any>();
+
+function stopTimeout(workspaceId: string) {
+  if (runningTimeouts.has(workspaceId)) {
+    clearTimeout(runningTimeouts.get(workspaceId));
+    runningTimeouts.delete(workspaceId);
+  }
+}
+
+/**
+ * @param minMs - Initial refresh delay
+ * @param maxMs - Maximum refresh delay
+ * @param stepFactor - Refresh delay multiplier
+ * @param limit - Maximum total time to keep refreshing
+ * @returns A generator that yields the next refresh delay
+ */
+function* refreshRangeIterator(
+  minMs: number,
+  maxMs: number,
+  stepFactor: number,
+  limit: number
+) {
+  let totalMs = 0;
+  do {
+    yield minMs;
+    totalMs += minMs;
+    minMs = Math.min(minMs * stepFactor, maxMs);
+  } while (totalMs < limit);
+}
+
+// Iterate with the delays until the predicate is true for the workspace
+function iterateUntilTrue(
+  iter: Generator<number>,
+  id: string,
+  predicate: () => boolean,
+  onDone?: () => void
+) {
+  const nextDelay = iter.next();
+  if (nextDelay.value && !predicate()) {
+    // Schedule the next iteration
+    log.debug("Scheduling next workspace refresh in ", nextDelay.value, "ms");
+    const timeoutId = setTimeout(
+      () => iterateUntilTrue(iter, id, predicate),
+      nextDelay.value
+    );
+    runningTimeouts.set(id, timeoutId);
+  } else {
+    // Either reached the time limit or the predicate is true and we're done
+    if (!nextDelay.value) {
+      log.debug("Reached time limit for workspace refresh");
+    } else {
+      log.debug("Predicate is true for workspace refresh. Done refreshing");
+    }
+    runningTimeouts.delete(id);
+    if (onDone) onDone();
+  }
+}
+
+export async function refreshUntilJobsAreFinished(
```

Comment:
Nit: since we never `await` this (as we shouldn't), and this doesn't `await` anything, the `Promise` is kind of worthless here... I think this doesn't need to be `async` and could be called something like `queueRefreshUntilJobsAreFinished` to convey the intent.